plugins/modules/pf.py

```diff
@@ -0,0 +1,320 @@
+    if state == 'started':
+        if not active_pf:
+            start_pf(module, active_pf)
+            changed = True
+
+    # Stop packet filter (pf) only it is currently running
+    if state == 'stopped':
+        if active_pf:
+            stop_pf(module, active_pf)
+            changed = True
+
+    # Restart of packet filter (pf) can always be performed
+    if state == 'restarted':
+        restart_pf(module, active_pf)
+        changed = True
+
+    # Reload/Flush the defined rule set for packet filter (pf)
+    # Note: May optionally run in dry run mode and print out
+    #       the rule set in Ansible meta output
+    if state == 'reloaded':
+        out = reload_pf(module, config, dry_run, filter)
+        # Dry run will not modify the system
+        if not dry_run:
+            changed = True
+
+    module.exit_json(
+        changed=changed,
+        meta={
+            "state": state,
+            "filter": filter,
+            "rule_set": out
+        }
+    )
+
+
+def validate(module, config):
+    """ Run basic validations """
+    _validate_os(module)
+    _validate_pf(module)
+    _validate_pf_config(module, config)
+
+
+def _validate_os(module):
+    """ Validate the operating system """
+    rc, out, err = module.run_command(['cat', '/etc/os-release'])
+
+    # Validate for a BSD string in output
+    if 'BSD' not in out:
+        msg_err = 'Error: Unsupported OS. This can only be used on BSD systems.'
+        module.fail_json(msg=msg_err)
+
+
+def _validate_pf(module):
+    """ Validate packet filter (pf) """
+    rc, out, err = module.run_command(['ls', '/sbin/pfctl'])
+
+    # Validate exit code
+    if rc != 0:
+        msg_err = 'Error: Unable to find pfctl binary.'
+        module.fail_json(msg=msg_err)
+
+
+def _validate_pf_config(module, config):
+    """ Validate if the defined config is present """
+    rc, out, err = module.run_command(['ls', config])
+
+    # Fail if no config file is present
+    if rc != 0:
+        msg_err = f'Error: Config file does not exist: {config}'
+        module.fail_json(msg=msg_err)
+
+
+def status_pf(module):
+    """ Current status of pf """
+    rc, out, err = module.run_command(['service', 'pf', 'status'])
+
+    # Obtain current status of pf
+    if 'Enabled' in out:
+        return True
+    else:
+        return False
+
+
+def start_pf(module, active_pf):
+    """ Start packet filter (pf) if not already running """
+    exec_opt = 'start'
+    error = False
+
+    rc, out, err = module.run_command(['service', 'pf', exec_opt])
+
+    # Validate exit code
+    if rc != 0:
+        error = True
+
+    # Validate for status change to make sure the desired
+    # state has been reached
+    new_active_pf = status_pf(module)
+    if new_active_pf == active_pf:
+        error = True
+
+    # Exit module on failure of change
+    if error:
+        msg_err = f'Error: Could not {exec_opt} pf.'
+        module.fail_json(msg=msg_err)
```

Comment:
This is effectively _squashing_ the details of the error. When that fails, the user will only know that something went bad but no useful information will be provided to them.

I would suggest using `check_rc=True` in the `run_command()` calls (here and throughout the code), and in line 220 (when checking whether the state was actually changed), to call `fail_json()` with something like `The pf service has not started` or similar.